internal/experiment/openvpn/openvpn_test.go

```diff
@@ -0,0 +1,421 @@
+			{Status: model.ArchivalOpenVPNConnectStatus{Success: true}},
+		}
+		if tk.AllConnectionsSuccessful() != false {
+			t.Fatal("expected false")
+		}
+	})
+	t.Run("all failures", func(t *testing.T) {
+		tk := openvpn.NewTestKeys()
+		tk.OpenVPNHandshake = []*model.ArchivalOpenVPNHandshakeResult{
+			{Status: model.ArchivalOpenVPNConnectStatus{Success: false}},
+			{Status: model.ArchivalOpenVPNConnectStatus{Success: false}},
+			{Status: model.ArchivalOpenVPNConnectStatus{Success: false}},
+		}
+		if tk.AllConnectionsSuccessful() != false {
+			t.Fatal("expected false")
+		}
+	})
+}
+
+func TestBadInputFailure(t *testing.T) {
+	m := openvpn.NewExperimentMeasurer(openvpn.Config{}, "openvpn")
+	ctx := context.Background()
+	sess := makeMockSession()
+	callbacks := model.NewPrinterCallbacks(sess.Logger())
+	measurement := new(model.Measurement)
+	measurement.Input = "openvpn://badprovider/?address=aa"
+	args := &model.ExperimentArgs{
+		Callbacks:   callbacks,
+		Measurement: measurement,
+		Session:     sess,
+	}
+	err := m.Run(ctx, args)
+	if !errors.Is(err, openvpn.ErrInvalidInput) {
+		t.Fatalf("expected ErrInvalidInput, got %v", err)
+	}
+}
+
+func TestVPNInput(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skip test in short mode")
+	}
+	// TODO -- do a real test, get credentials etc.
+}
+
+func TestMeasurer_FetchProviderCredentials(t *testing.T) {
+	t.Run("Measurer.FetchProviderCredentials calls method in session", func(t *testing.T) {
+		m := openvpn.NewExperimentMeasurer(
+			openvpn.Config{},
+			"openvpn").(openvpn.Measurer)
+
+		sess := makeMockSession()
+		_, err := m.FetchProviderCredentials(
+			context.Background(),
+			sess, "riseup")
+		if err != nil {
+			t.Fatal("expected no error")
+		}
+	})
+	t.Run("Measurer.FetchProviderCredentials raises error if API calls fail", func(t *testing.T) {
+		someError := errors.New("unexpected")
+
+		m := openvpn.NewExperimentMeasurer(
+			openvpn.Config{},
+			"openvpn").(openvpn.Measurer)
+
+		sess := makeMockSession()
+		sess.MockFetchOpenVPNConfig = func(context.Context, string, string) (*model.OOAPIVPNProviderConfig, error) {
+			return nil, someError
+		}
+		_, err := m.FetchProviderCredentials(
+			context.Background(),
+			sess, "riseup")
+		if !errors.Is(err, someError) {
+			t.Fatalf("expected error %v, got %v", someError, err)
+		}
+	})
+
+}
+
+func TestSuccess(t *testing.T) {
+	m := openvpn.NewExperimentMeasurer(openvpn.Config{
+		Provider: "riseup",
+		SafeCA:   "base64:Zm9v",
+		SafeKey:  "base64:Zm9v",
+		SafeCert: "base64:Zm9v",
+	}, "openvpn")
+	ctx := context.Background()
+	sess := makeMockSession()
+	callbacks := model.NewPrinterCallbacks(sess.Logger())
+	measurement := new(model.Measurement)
+	measurement.Input = "openvpn://riseupvpn.corp/?address=127.0.0.1:9989&transport=tcp"
+	args := &model.ExperimentArgs{
+		Callbacks:   callbacks,
+		Measurement: measurement,
+		Session:     sess,
+	}
+	if sess.Logger() == nil {
+		t.Fatal("logger should not be nil")
+	}
```

Comment:
Why are we testing for logger to be not-nil here?

Reply:
no idea, but does not seem relevant to test here.